### ISSUE TITLE
Rotate non-CJK glyphs embedded in vertically-oriented CJK labels

### ISSUE THREAD

**Issue by 1ec5** [mapbox/mapbox-gl-js#3506]
Per https://github.com/mapbox/mapbox-gl-js/issues/1246#issuecomment-255554493, when a label is vertically oriented with a vertical writing mode, any runs of non-CJK characters (characters that would normally be oriented horizontally) should be rotated as if the label were oriented using a horizontal writing mode.

In other words, once #3438 lands, non-CJK glyphs will remain upright when in vertically oriented labels:

![](https://www.w3.org/TR/jlreq/Images-en/img2_2_5.png)

Instead, these non-CJK glyphs should be rotated as they would’ve been before #3438:

![](https://www.w3.org/TR/jlreq/Images-en/img2_2_6.png)

On OpenStreetMap, and therefore in Mapbox Streets, mixed CJK/romanized labels are ubiquitous on roads in [Hong Kong](http://www.openstreetmap.org/way/45740459), [Macao](http://www.openstreetmap.org/way/164166938), and [Shenzhen](http://www.openstreetmap.org/way/269895566). For example, [this way](http://www.openstreetmap.org/way/45740459) is tagged `name=海底隧道 Cross-Harbour Tunnel`. Unless we take care to rotate the Latin glyphs and advance them by the normal horizontal advance (rather than the glyph height), the label will become much longer when oriented vertically and stand a much higher chance of getting collided out.

----

There’s one more detail to be aware of: one- or two-digit numbers (and one- or two-letter words) would ideally be left upright, their digits placed side-by-side. This effect is called _tate-chu-yoko_ in Japanese:

![](https://www.w3.org/TR/jlreq/Images-en/img2_2_7.png)

In order for this to work, we’d need to replace ASCII letters and numbers with their fullwidth or halfwidth equivalents. I think _tate-chu-yoko_ would be a much lower priority than the suggestion above to rotate the glyphs.

----

The illustrations above come from the W3C’s “[Requirements for Japanese Text Layout](https://www.w3.org/TR/jlreq/#japanese_and_western_mixed_text_composition)” document, which covers a number of other details needed for proper typographic treatment of Japanese text. Many of these details also apply to Chinese. However, there’s a point of diminishing returns for us, because we’re dealing with map labels rather than magazine text layout.

/cc @lucaswoj @nickidlugash @xrwang @friedbunny

**Comment by 1ec5** [comment]
There may be some interplay between this issue and #3505, because we’d need to avoid rotating punctuation that lies within non-CJK text.

**Comment by lucaswoj** [comment]
Rather than mark this a "release blocker", I'm going to roll this into #3438

**Comment by lucaswoj** [comment]
Done in #3438

**Comment by 1ec5** [comment]
Next steps: #3588.
